### PR TITLE
add FeaStConv

### PR DESCRIPTION
**Pull request by sw-gong**
Please check if there is any issue. https://arxiv.org/pdf/1706.05206.pdf

I used translation invariant assignments which reported to have better performance instead of the formula (2).

### CODE REVIEW

**Comment by sw-gong** [comment]
I don't understand what these two codecov fails mean. Can you please explain about this? Thank you.

**Comment by rusty1s** [comment]
Test Coverage decreased.

**Comment by rusty1s** [comment]
If you do not have time, I can write a test for it.

**Comment by sw-gong** [comment]
It finally passed. I guess I will know how to give a good PR next time. :-) Btw, please check the implementation with the paper. Thanks.

**Comment by rusty1s** [comment]
For linting, I do autoformatting with`yapf`, and run `python setup.py test` before comitting. We should probably create a `CONTRIBUTE.md` for all other contributors.

**Comment by rusty1s** [comment]
Will check the code by tomorrow. Thanks again for your awesome contributions :)

**Comment by sw-gong** [comment]
> For linting, I do autoformatting with`yapf`, and run `python setup.py test` before comitting. We should probably create a `CONTRIBUTE.md` for all other contributors.

Thank you. I'll have a look it.

**Comment by rusty1s** [comment]
Hi, just some minor questions regarding this line: `q = torch.mm((x_i - x_j), self.u) + self.c`.

The paper states that it computes `q` by two parameters ` u` and `v` instead of just one. In addition, we do not really need `matmul` here, a basic `(x_j * self.u).sum(dim=1)` should be way faster (as far as I know).

**Comment by sw-gong** [comment]
> Hi, just some minor questions regarding this line: `q = torch.mm((x_i - x_j), self.u) + self.c`.
> 
> The paper states that it computes `q` by two parameters ` u` and `v` instead of just one. In addition, we do not really need `matmul` here, a basic `(x_j * self.u).sum(dim=1)` should be way faster (as far as I know).

Hi, as I said at the beginning I used  translation invariant assignments which reported to have better performance instead of the formula (2) in the paper. As for using dot product replacing torch.matmul, I agree with you. 

image
image

**Comment by rusty1s** [comment]
Ah, I'm sorry, you are right. I should read more closely :P

**Comment by sw-gong** [comment]
No worries. :) 

Below is the repository posted by the paper author. I'm not familiar with tf but maybe you can have a look at it if you want. I ran experiments on FAUST with the same architecture shown in the paper (don't know hyperparameter setup), the accuracy of which is around 8-10% lower than reported 92.2%. 
https://github.com/nitika-verma/FeaStNet

**Comment by dhorka** [comment]
I have one doubt related to this implementation. I was checking the paper I do not see where they use attention heads and how many they are using.

**Comment by rusty1s** [comment]
Well, this denotes the number of weight matrices the convolution will use. A more appropriate name would be indeed `kernel_size`. I can change the naming. WDYT @sw-gong?

**Comment by dhorka** [comment]
Then the output of the layer it will be weight_matrices*out_channels? I'm a bit confusing about that.

**Comment by rusty1s** [comment]
No, the output will be `[num_nodes, out_channels]`. We sum up the results of all weight matrices, see [here](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.FeaStConv).

**Comment by dhorka** [comment]
Then maybe it is not appropriate kernel_size, right? Normally kernel_size is used to define the filter of a CNN, in this case we have multiple filters. Maybe the nomenclature of the paper can be preserved and use M.

**Comment by rusty1s** [comment]
If I am not totally mistaken, `kernel_size` is the appropriate name. In a traditional CNN, you also have multiple filters, with each kernel point having `[in_channels, out_channels]` weights. This is also discussed in the paper:
Screenshot 2019-07-22 at 09 44 43
